### PR TITLE
[Metrics] Added additional metrics for CASE and Subscription

### PR DESCRIPTION
**Pull request by anush-apple**
- CASE session tracking
- Subscription Setup
- MRP Retries

### CODE REVIEW

File: src/app/ReadClient.cpp
Comment (tcarmelveilleux) [inline src/app/ReadClient.cpp:192]: Same event recorded in two places under different conditions may dilute the understanding of the underlying reason.

File: src/protocols/secure_channel/CASESession.cpp
Comment (tcarmelveilleux) [inline src/protocols/secure_channel/CASESession.cpp:1330]: `CHIP_ERROR err = .....`

File: src/protocols/secure_channel/CASESession.cpp
Comment (tcarmelveilleux) [inline src/protocols/secure_channel/CASESession.cpp:1331]: Curuous why this is `Sigma1` related event rather than Sigma2

**Review by tcarmelveilleux — APPROVED** [review]

File: src/app/ReadClient.cpp
Comment (anush-apple) [reply src/app/ReadClient.cpp:192]: I was trying to capture all cases when subscription would fail. Close seems liked the funnel point. Would we not want to capture if closed due to initial report or waiting for a response?

File: src/protocols/secure_channel/CASESession.cpp
Comment (anush-apple) [reply src/protocols/secure_channel/CASESession.cpp:1330]: Will fix

File: src/protocols/secure_channel/CASESession.cpp
Comment (anush-apple) [reply src/protocols/secure_channel/CASESession.cpp:1331]: This is marking the end of Sigma1 from the initiator side once it receives the Sigma2 from the responder.

File: src/app/ReadClient.cpp
Comment (bzbarsky-apple) [inline src/app/ReadClient.cpp:500]: If something inside ProcessSubscribeResponse() before it calls `MoveToState` fails, then we will do the MATTER_LOG_METRIC_END here _and_ in the resulting Close() call, right?

I wonder whether it would be better to put ` MATTER_LOG_METRIC_END(Tracing::kMetricDeviceSubscriptionSetup, CHIP_NO_ERROR);` in either MoveToState when moving to the `SubscriptionActive` state or at the one place where we move to that state.  That way we unambiguously do the "ended with error" in Close() and "ended with success" on the codepath that does that.

File: src/app/ReadClient.cpp
Comment (bzbarsky-apple) [inline src/app/ReadClient.cpp:1117]: I assume we don't care that the unit-test-only `SendSubscribeRequestWithoutValidation` calls `SendSubscribeRequestImpl` and then does not do the "end on failure" behavior that `SendSubscribeRequest` was modified to do?  If that's the case, might be worth documenting.  Or we could try to just fix both of the callsites to have identical behavior.

It would be nice, perhaps, if we had a macro for "do start here, do end if err is failure when we are returning" of some sort....  But that's out of scope for this PR.

File: src/darwin/Framework/CHIP/MTRDevice.mm
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDevice.mm:119]:
```suggestion

typedef void (^FirstReportHandler)(void);
```

File: src/darwin/Framework/CHIP/MTRDevice.mm
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDevice.mm:1010]: So this will get called on internal re-subscribes from the ReadClient, so we can get a MATTER_LOG_METRIC_END here without a corresponding START.  Is that OK?  If not, consider conditioning this on whether `HadSubscriptionEstablishedOnce()` is true (see code a bit below this).

File: src/darwin/Framework/CHIP/MTRDevice.mm
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDevice.mm:1060]: So this function is called in three cases:

1. Getting a CASE session for the initial subscription fails.
2. Kicking off the subscription on the CASE session we got fails.
3. We failed to kick off a resubscribe for some reason.

Item 3 does not seem relevant to us, at first glance.  Maybe this should be inside `_handleSubscriptionReset`, though that can also get called if everything here is being torn down.....

On the other hand, how do we want to handle the case of "got a CASE session, but first attempt to set up a subscription failed, so we will try again"?  Right now we will keep waiting for it and eventually when it's established log a "success" end metric.  Is that what we want?

Also: we seem to be treating subscription reattempts that come from MTRDevice (i.e. "before we get a CASE session") and the ones that come from inside ReadClient (i.e. "after we get a CASE session") differently: the former log begin/end on every attempt, while the latter logs begin on first attempt start and then end on success, which might be a while later).  This asymmetry is a bit strange to me.

**Review by woody-apple — APPROVED** [review]
Approving pending comments

File: src/app/ReadClient.cpp
Comment (anush-apple) [reply src/app/ReadClient.cpp:500]: The backend needs to handle multiple end or begin calls. The usual way is to capture the first of these events.

File: src/darwin/Framework/CHIP/MTRMetricKeys.h
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRMetricKeys.h:91]:
```suggestion
// Setup from darwin MTRDevice for initial subscription to a device
```

File: src/app/ReadClient.cpp
Comment (anush-apple) [reply src/app/ReadClient.cpp:192]: The handler of the _BEGIN and _END  usually needs to account for multiple invocations. Usually, the first instance of the event is captured and the remaining ones are ignored. Given that, this should be fine since we are capturing the failure only if the subscription setup fails.

**Pull request merged**
